### PR TITLE
[serve] Lazily construct `asyncio.Event` to avoid attaching to the wrong loop

### LINKED ISSUE #40631: [serve] `asyncio.Event` in router attached to wrong loop
Spotted in some test logs:
```
tests/test_healthcheck.py::test_no_user_defined_method[True] 2023-10-24 12:03:20,104    INFO worker.py:1659 -- Started a local Ray instance. View the dashboard at http://127.0.0.1:8265
(ServeController pid=11210) INFO 2023-10-24 12:03:22,297 controller 11210 deployment_state.py:1379 - Deploying new version of deployment A in application 'default'.
(ProxyActor pid=11213) INFO 2023-10-24 12:03:22,257 proxy 127.0.0.1 proxy.py:1100 - Proxy actor 12d894e6989f447618ef5b2401000000 starting on node eab120fa3ee99e3df5649cdc3d3a9edf1772294da7b29fdce52b7b13.
(ProxyActor pid=11213) INFO 2023-10-24 12:03:22,261 proxy 127.0.0.1 proxy.py:1285 - Starting HTTP server on node: eab120fa3ee99e3df5649cdc3d3a9edf1772294da7b29fdce52b7b13 listening on port 8000
(ProxyActor pid=11213) INFO:     Started server process [11213]
(ServeController pid=11210) INFO 2023-10-24 12:03:22,398 controller 11210 deployment_state.py:1668 - Adding 1 replica to deployment A in application 'default'.
2023-10-24 12:03:24,304 INFO router.py:910 -- Using router <class 'ray.serve._private.router.PowerOfTwoChoicesReplicaScheduler'>.
2023-10-24 12:03:24,314 INFO router.py:470 -- Got updated replicas for deployment 'A' in application 'default': {'default#A#qdirvY'}.
2023-10-24 12:03:24,323 WARNING router.py:724 -- Failed to fetch queue length for replica default#A#qdirvY: 'The actor died unexpectedly before finishing this task.
        class_name: ServeReplica:default:A
        actor_id: 7185eb52627c74b77fc34b3d01000000
        pid: 11216
        name: SERVE_REPLICA::default#A#qdirvY
        namespace: serve
        ip: 127.0.0.1
The actor is dead because it was killed by `ray.kill`.' This replica will no longer be considered for requests.
2023-10-24 12:03:24,323 INFO router.py:562 -- Tried to assign replica for deployment 'A' in application 'default' but none are available. Waiting for new replicas to be added.
2023-10-24 12:03:24,323 ERROR router.py:815 -- Unexpected error in fulfill_pending_requests.
Traceback (most recent call last):
  File "/Users/eoakes/code/ray/python/ray/serve/_private/router.py", line 806, in fulfill_pending_requests
    async for candidates in self.choose_two_replicas_with_backoff(
  File "/Users/eoakes/code/ray/python/ray/serve/_private/router.py", line 569, in choose_two_replicas_with_backoff
    await self._replicas_updated_event.wait()
  File "/Users/eoakes/miniforge3/envs/ray/lib/python3.8/asyncio/locks.py", line 309, in wait
    await fut
RuntimeError: Task <Task pending name='Task-6' coro=<PowerOfTwoChoicesReplicaScheduler.fulfill_pending_requests() running at /Users/eoakes/code/ray/python/ray/serve/_private/router.py:806>> got Future <Future pending> attached to a different loop
```

I'm not 100% sure how to repro or why this isn't causing test failures yet, looking into it.

### PR DESCRIPTION
**Pull request by edoakes**
## Why are these changes needed?

Python 3.10 removed the `loop` parameter to `asyncio.Event`. Now, the `asyncio.Event` will call `get_running_loop` in its constructor to determine the loop to attach to. This class can be constructed for the handle from a different loop than it uses for scheduling, so we need to construct it lazily to avoid an error due to the event being attached to the wrong loop.

## Related issue number

Closes https://github.com/ray-project/ray/issues/40631

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/_private/router.py [hunk 1]
```diff
@@ -344,7 +344,14 @@ def __init__(
...
         self._replica_id_set: Set[str] = set()
         self._replicas: Dict[str, ReplicaWrapper] = {}
-        self._replicas_updated_event = asyncio.Event()
+
+        # NOTE(edoakes): Python 3.10 removed the `loop` parameter to `asyncio.Event`.
+        # Now, the `asyncio.Event` will call `get_running_loop` in its constructor to
+        # determine the loop to attach to. This class can be constructed for the handle
+        # from a different loop than it uses for scheduling, so we need to construct it
+        # lazily to avoid an error due to the event being attached to the wrong loop.
+        self._lazily_initialied_replicas_updated_event: Optional[asyncio.Event] = None
```
Comment (zcin) [inline python/ray/serve/_private/router.py:353]:
```suggestion
        self._lazily_initialized_replicas_updated_event: Optional[asyncio.Event] = None
```

**Review by zcin — COMMENTED** [review]
Is it possible/necessary to add a test for the case you mentioned (the class being constructed from a different loop)?

File: python/ray/serve/_private/router.py (hunk 1 above)
Comment (edoakes) [reply python/ray/serve/_private/router.py:353]: Was theory-crafting how to add a test, just did, PTAL!

**Review by GeneDer — APPROVED** [review]
LGTM!

**Review by shrekris-anyscale — APPROVED** [review]
Nice!

**Pull request merged**